components/eamxx/src/dynamics/homme/eamxx_homme_iop.cpp

```diff
@@ -0,0 +1,408 @@
+#include "eamxx_homme_process_interface.hpp"
+
+// EAMxx includes
+#include "control/intensive_observation_period.hpp"
+#include "dynamics/homme/homme_dimensions.hpp"
+#include "dynamics/homme/homme_dynamics_helpers.hpp"
+#include "physics/share/physics_constants.hpp"
+#include "share/util/scream_column_ops.hpp"
+
+// Homme includes
+#include "Context.hpp"
+#include "ColumnOps.hpp"
+#include "ElementOps.hpp"
+#include "EquationOfState.hpp"
+#include "HommexxEnums.hpp"
+#include "HybridVCoord.hpp"
+#include "KernelVariables.hpp"
+#include "SimulationParams.hpp"
+#include "Types.hpp"
+
+// EKAT includes
+#include "ekat/ekat_workspace.hpp"
+#include "ekat/kokkos/ekat_kokkos_types.hpp"
+
+namespace scream {
+
+// Compute effects of large scale subsidence on T, q, u, and v.
+KOKKOS_FUNCTION
+void HommeDynamics::
+advance_iop_subsidence(const KT::MemberType& team,
+                       const int nlevs,
+                       const Real dt,
+                       const Real ps,
+                       const view_1d<const Pack>& pmid,
+                       const view_1d<const Pack>& pint,
+                       const view_1d<const Pack>& pdel,
+                       const view_1d<const Pack>& omega,
+                       const Workspace& workspace,
+                       const view_1d<Pack>& u,
+                       const view_1d<Pack>& v,
+                       const view_1d<Pack>& T,
+                       const view_2d<Pack>& Q)
+{
+  using ColOps = ColumnOps<DefaultDevice, Real>;
+  using C = physics::Constants<Real>;
+  constexpr Real Rair = C::Rair;
+  constexpr Real Cpair = C::Cpair;
+
+  const auto n_q_tracers = Q.extent_int(0);
+  const auto nlev_packs = ekat::npack<Pack>(nlevs);
+
+  // Get some temporary views from WS
+  uview_1d<Pack> omega_int, delta_u, delta_v, delta_T, tmp;
+  workspace.take_many_contiguous_unsafe<4>({"omega_int", "delta_u", "delta_v", "delta_T"},
+                                           {&omega_int,  &delta_u,  &delta_v,  &delta_T});
+  const auto delta_Q_slot = workspace.take_macro_block("delta_Q", n_q_tracers);
+  uview_2d<Pack> delta_Q(delta_Q_slot.data(), n_q_tracers, nlev_packs);
+
+  auto s_pmid = ekat::scalarize(pmid);
+  auto s_omega = ekat::scalarize(omega);
+  auto s_delta_u = ekat::scalarize(delta_u);
+  auto s_delta_v = ekat::scalarize(delta_v);
+  auto s_delta_T = ekat::scalarize(delta_T);
+  auto s_delta_Q = ekat::scalarize(delta_Q);
+  auto s_omega_int = ekat::scalarize(omega_int);
+
+  // Compute omega on the interface grid by using a weighted average in pressure
+  const int pack_begin = 1/Pack::n, pack_end = (nlevs-1)/Pack::n;
+  Kokkos::parallel_for(Kokkos::TeamVectorRange(team, pack_begin, pack_end+1), [&] (const int k){
+    auto range_pack = ekat::range<IntPack>(k*Pack::n);
+    range_pack.set(range_pack<1, 1);
+    Pack pmid_k, pmid_km1, omega_k, omega_km1;
```

Comment:
Would this be legal to move Pack definition outside of TeamVectorRange loop? The only reason I question is in https://kokkos.org/kokkos-core-wiki/ProgrammingGuide/HierarchicalParallelism.html#team-loops 

> Note that it is not legal to make a write access to POD data outside the closure of a nested parallel layer. This is a conscious choice to prevent difficult-to-debug issues related to thread private, team shared and globally shared variables. A simple way to enforce this is by using the “capture by value”’ clause with lambdas, but “capture by reference” is recommended for release builds...

I guess `Pack` isn't exactly a POD type, but capturing by value would fail.

Testing though, for openMP it looks like it works and give identical answers.

Reply:
Uhm, that may be a valid point. I need to double check this with the kokkos folks.

Reply:
It appears you are correct. The temp needs to be declared locally.

Reply:
How significant is the performance hit?